File: Medium/113.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
    result = []
    path = []

    def dfs(root, targetSum):
      if root is None:
        return

      path.append(root.val)
      targetSum -= root.val
      
      # debug
      print("current path:", path, "remaining:", targetSum)

      if root.left is None and root.right is None and targetSum == 0:
        result.append(path[:])

      dfs(root.left, targetSum)
      dfs(root.right, targetSum)

      path.pop()

    dfs(root, targetSum)
    return result
  
  
  # 手动测试

root = TreeNode(5)
root.left = TreeNode(4)
root.right = TreeNode(8)
root.left.left = TreeNode(11)
root.left.left.left = TreeNode(7)
root.left.left.right = TreeNode(2)
root.right.left = TreeNode(13)
root.right.right = TreeNode(4)
root.right.right.left = TreeNode(5)
root.right.right.right = TreeNode(1)
solution = Solution()

print(solution.pathSum(root, 22))

File: Medium/113.cs
using System;
using System.Collections.Generic;

// Definition for a binary tree node.
// public class TreeNode {
//   public int val;
//   public TreeNode left;
//   public TreeNode right;
//   public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
//     this.val = val;
//     this.left = left;
//     this.right = right;
//   }
// }

public class Solution {
  public IList<IList<int>> PathSum(TreeNode root, int targetSum) {
    IList<IList<int>> result = new List<IList<int>>();
    List<int> path = new List<int>();

    void Dfs(TreeNode root, int targetSum) {
      if (root == null) {
        return;
      }

      path.Add(root.val);
      targetSum -= root.val;

      if (root.left == null && root.right == null && targetSum == 0) {
        result.Add(new List<int>(path));
      }

      Dfs(root.left, targetSum);
      Dfs(root.right, targetSum);

      path.RemoveAt(path.Count - 1);
    }

    Dfs(root, targetSum);
    return result;
  }
}